### PR TITLE
EES-2771 New feature toggle adding location hierarchies in Subject Meta Data

### PR DESCRIPTION
**Pull request by benoutram**
This PR introduces a new feature toggle which when enabled allows configuring location hierarchies in subject meta data.

A new response format is defined for the locations field in the subject meta data response (compare examples below of responses with the feature toggle turned off and on).

A similar change but for _table result_ subject meta data is also planned in EES-2881.

The feature toggle is intended to stay in place until the UI work for Location hierarchies is completed  by EES-2773 and EES-2777.

This new response format will be made permanent and the old format will be removed by EES-2902.

## Affected API calls

Admin App

* `GET /api/data/meta/subject/{subject_id}`
* `POST /api/data/meta/subject`

Data.Api App

* `GET /api/meta/subject/{subject_id}`
* `POST /api/meta/subject`

## Other changes

This PR begins to replace the term 'Observational Unit' with 'Location Attribute' used for the constituent parts of a Location, e.g. Country, Region, Local Authority, School, Provider, Planning Area, Ward etc. 

Although 'Observational Unit' may be a term that DfE analysts are familiar with it seems to have caused confusion in conversations around Locations in codebase as to what it refers to and the difference between an 'Observational Unit' and a 'Location'. Hopefully the term 'Location Attribute' removes this ambiguity.

A further update to completely remove the abstract 'ObservationalUnit' type is still required and some view models continue to use the term as a field name for now.

## Turning on/off

To toggle the hierarchy feature, in the AppSettings configuration of Admin and Data.API app services find the following configuration.

```
  "FeatureManagement": {
    "LocationHierarchies" : false
  },
```

The property `LocationHierarchies` can be toggled between `true` and `false`.

## Configuring Location hierarchies

Hierarchies are configured per level with the Location attributes that make up the hierarchy.

Example: For Local Authority level data where Country, Region and Local Authority attributes are provided,
a hierarchy Country -> Region -> Local Authority can be configured for this level using configuration data.

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Country",
        "Region",
        "LocalAuthority"
      ]
    }
  }
```

This alone won't create any hierarchy within Regional level data. For that, a hierarchy needs to be defined for the Region level.

A hierarchy Region -> Local Authority could be configured as follows:

```
  "Locations":
  {
    "Hierarchies": {
      "Region": [
        "Country",
        "Region"
      ]
    }
  }
```

It's possible to define hierarchies for multiple levels in the configuration e.g.

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Country",
        "Region",
        "LocalAuthority"
      ],
      "Region": [
        "Country",
        "Region"
      ],
      "Provider": [
        "ProviderType",
        "Provider"
      ],
      "School": [
        "Region",
        "LocalAuthority",
        "School"
      ]
    }
  }
```

Defining hierarchies does not change any behaviour when the feature toggle is turned off.

## Behaviour with feature toggle OFF

With the feature toggle turned off, subject meta data contains Locations without any changes to what we expect in a response prior to this change.

Here's an example where Country, Regional, and LA data is provided,  abbreviated to only include East Midlands Region and Derby and Nottingham LA's.

```
{
  "locations": {
    "country": {
      "hint": "",
      "legend": "National",
      "options": [
        {
          "label": "England",
          "value": "E92000001"
        }
      ]
    },
    "region": {
      "hint": "",
      "legend": "Regional",
      "options": [
        {
          "label": "East Midlands",
          "value": "E12000004"
        }
      ]
    },
    "localAuthority": {
      "hint": "",
      "legend": "Local Authority",
      "options": [
        {
          "label": "Derby",
          "value": "E06000015"
        },
        {
          "label": "Nottingham",
          "value": "E06000018"
        }
      ]
    }
  }
}
```

## Behaviour with feature toggle ON

With the feature toggle turned off, subject meta data contains Locations grouped by any hierarchies that are defined in configuration.

All examples use data where Country, Regional, and LA data is provided,  abbreviated to only include East Midlands Region and Derby and Nottingham LA's.

#### Configuration

```
  "Locations":
  {
    "Hierarchies": {
    }
  }
```

#### Response

```
{
  "locations": {
    "country": {
      "legend": "National",
      "options": [
        {
          "label": "England",
          "value": "E92000001"
        }
      ]
    },
    "region": {
      "legend": "Regional",
      "options": [
        {
          "label": "East Midlands",
          "value": "E12000004"
        }
      ]
    },
    "localAuthority": {
      "legend": "Local Authority",
      "options": [
        {
          "label": "Derby",
          "value": "E06000015"
        },
        {
          "label": "Nottingham",
          "value": "E06000018"
        }
      ]
    }
  }
}
```

#### Configuration

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Region",
        "LocalAuthority"
      ]
    }
  }
```

#### Response

```
{
  "locations": {
    "country": {
      "legend": "National",
      "options": [
        {
          "label": "England",
          "value": "E92000001"
        }
      ]
    },
    "region": {
      "legend": "Regional",
      "options": [
        {
          "label": "East Midlands",
          "value": "E12000004"
        }
      ]
    },
    "localAuthority": {
      "legend": "Local Authority",
      "options": [
        {
          "label": "East Midlands",
          "value": "E12000004",
          "level": "Region",
          "options": [
            {
              "label": "Derby",
              "value": "E06000015"
            },
            {
              "label": "Nottingham",
              "value": "E06000018"
            }
          ]
        }
      ]
    }
  }
}
```

#### Configuration

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Country",
        "Region",
        "LocalAuthority"
      ]
    }
  }
```

#### Response

```
{
  "locations": {
    "country": {
      "legend": "National",
      "options": [
        {
          "label": "England",
          "value": "E92000001"
        }
      ]
    },
    "region": {
      "legend": "Regional",
      "options": [
        {
          "label": "East Midlands",
          "value": "E12000004"
        }
      ]
    },
    "localAuthority": {
      "legend": "Local Authority",
      "options": [
        {
          "label": "England",
          "value": "E92000001",
          "level": "Country",
          "options": [
            {
              "label": "East Midlands",
              "value": "E12000004",
              "level": "Region",
              "options": [
                {
                  "label": "Derby",
                  "value": "E06000015"
                },
                {
                  "label": "Nottingham",
                  "value": "E06000018"
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

#### Configuration

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Country",
        "Region",
        "LocalAuthority"
      ],
      "Region": [
        "Country",
        "Region"
      ]
    }
  }
```

#### Response

```
{
  "locations": {
    "country": {
      "legend": "National",
      "options": [
        {
          "label": "England",
          "value": "E92000001"
        }
      ]
    },
    "region": {
      "legend": "Regional",
      "options": [
        {
          "label": "England",
          "value": "E92000001",
          "level": "Country",
          "options": [
            {
              "label": "East Midlands",
              "value": "E12000004"
            }
          ]
        }
      ]
    },
    "localAuthority": {
      "legend": "Local Authority",
      "options": [
        {
          "label": "England",
          "value": "E92000001",
          "level": "Country",
          "options": [
            {
              "label": "East Midlands",
              "value": "E12000004",
              "level": "Region",
              "options": [
                {
                  "label": "Derby",
                  "value": "E06000015"
                },
                {
                  "label": "Nottingham",
                  "value": "E06000018"
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

### Example scenario where all Observations have Locations that don't have an attribute present in the hierarchy.

E.g a Country-Region-LA hierarchy is defined for the LA level but no Locations are provided with a Region. A data file would look something like this:

```
geographic_level    country_code    country_name    region_code    region_name    la_code    la_name
====================================================================================================
Local authority     E92000001       England                                       E06000015  Derby
Local authority     E92000001       England                                       E06000015  Derby
Local authority     E92000001       England                                       E06000018  Nottingham
```

#### Configuration

```
  "Locations":
  {
    "Hierarchies": {
      "LocalAuthority": [
        "Country",
        "Region",
        "LocalAuthority"
      ]
    }
  }
```

#### Response

```
{
  "locations": {
    "localAuthority": {
      "legend": "Local Authority",
      "options": [
        {
          "label": "England",
          "value": "E92000001",
          "level": "Country",
          "options": [
            {
              "label": "",
              "value": "",
              "level": "Region",
              "options": [
                {
                  "label": "Derby",
                  "value": "E06000015"
                },
                {
                  "label": "Nottingham",
                  "value": "E06000018"
                }
              ]
            }
          ]
        }
      ]
    }
  }
}
```

image

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Data.Api/appsettings.Development.json [hunk 1]
```diff
@@ -13,6 +13,9 @@
       "Microsoft.EntityFrameworkCore.Database.Command": "Warning"
     }
   },
+  "FeatureManagement": {
+    "LocationHierarchies" : false
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Api/appsettings.Development.json:17]: I can't seem to see where this is causing features to be switched on and off?  Was expecting to see some mention of "LocationHierarchies" in the codebase other than in these files?

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs [hunk 2]
```diff
@@ -1,14 +1,15 @@
+#nullable enable
 namespace GovUk.Education.ExploreEducationStatistics.Data.Model
 {
-    public class Country : ObservationalUnit
+    public class Country : ObservationalUnit, ILocationAttribute
     {
-        public Country(string code, string name) : base(code, name)
+        public Country(string? code, string? name) : base(code, name)
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs:6]: What's the driver behind making these nullable now?

File: src/GovUk.Education.ExploreEducationStatistics.Data.Api/appsettings.Development.json (hunk 1 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Api/appsettings.Development.json:17]: Oh wait, sorry I see it!

Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Api/appsettings.Development.json:17]: It's used in `SubjectMetaService.cs`. Check for `_featureManager.IsEnabledAsync`

File: src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json [hunk 3]
```diff
@@ -37,6 +37,9 @@
       "MinutesBeforeReleaseTimeEnd": 1
     }
   },
+  "FeatureManagement": {
```
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json:40]: Does omitting this setting in other environments set it to false by default?

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs (hunk 2 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs:6]: There should be no overall change here and I'm not trying to rework anything except to represent their current nullable state correctly in the model after setting `Location` and the attribute types (e.g. Country) as nullable reference types with `nullable enable`. 

In the example of Country, Location `Country_Code` and `Country_Name` fields are nullable in the database so they have now been marked as nullable fields in the `Location` type.

You can always 'get' Location.Country but if the values are not provided then it's constructed with a null code and name so these constructor params are nullable too.

If I don't make this change to the `Location` type I think I would get a generated migration altering all the fields of the table to be not null. I have checked to make sure that no migration code is generated with these changes.

If I make the change to the `Location` type but didn't make this change to the attribute types (e.g. the change to `Country` here) then we get code analysis warnings.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json (hunk 3 above)
Comment (benoutram) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json:40]: Yes, it should be off by default.

I doubt that we would ever want to turn it on in any of the other environments as the idea behind it is that we can get this change into the codebase to unblock further development work (i.e. hierarchies within table result subject meta data, ordering of hierarchies, and the front end support), without breaking what we have currently or avoid a long running branch hanging around that all further work depends on.

I would think that as soon as the front end work is done we will remove this setting as I don't think there's any plan for the front end to support the old non-hierarchical format, at least for subject meta data.

I could have just made this toggle depend on a boolean property of the `LocationOptions` class instead but I was experimenting with this dependency to see what it gives us. I suppose we could make use of it in future if we want to keep it registered. 

There should also be somewhere in the App Service UI in Azure to toggle this setting, although without deploying it I don't know what that looks like.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs (hunk 2 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Model/Country.cs:6]: Understood, thanks!

File: src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json (hunk 3 above)
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/appsettings.Development.json:40]: Nice one, thanks

**Review by duncan-at-hiveit — APPROVED** [review]

**Pull request merged**